syntax-highlighting-code-block.php

```diff
@@ -203,13 +222,28 @@ function render_block( $attributes, $content ) {
 			$r = $highlighter->highlightAuto( $code );
 		}
 
-		$code        = $r->value;
-		$language    = $r->language;
-		$highlighted = compact( 'code', 'language' );
+		$code       = $r->value;
+		$language   = $r->language;
+		$show_lines = $attributes['showLines'];
+
+		if ( $show_lines ) {
+			require_once __DIR__ . '/vendor/scrivo/highlight.php/HighlightUtilities/functions.php';
+
+			$lines = \HighlightUtilities\splitCodeIntoArray( $code );
+			$code  = '';
+
+			// We need to wrap the line of code twice in order to let out `white-space: pre` CSS setting to be respected
+			// by our `table-row`.
+			foreach ( $lines as $line ) {
+				$code .= sprintf( '<div class="loc"><span>%s</span></div>%s', $line, PHP_EOL );
```

Comment:
In the future it will be useful to include the `loc` even if line numbers aren't being displayed. This will make it easy to highlight lines.

Reply:
Are you okay with adding the extra overhead on the initial render of splitting up code into lines even if it's not showing line numbers or highlighting lines?

Reply:
I think so. How much overhead is it?

Reply:
It's nothing too crazy:

In `splitCodeIntoArray()` from highlight.php:

1. It'll parse the highlighted HTML into a `\DOMDocument`
2. Loop through all of the spans and see if there are any new lines
   1. If new lines exist, modify the lines of code to add closing and opening tags
3. Split the modified code into an array.

In this plug-in:

1. Loop through all of the lines of code and wrap them.

Reply:
I wonder. Why is `splitCodeIntoArray` even needing to load the string into a `DOMDocument` to do this? Why can't it just split the string by lines given that it is inside of a `pre`?

Reply:
Parsing it as a DOM is necessary, take this snippet as an example:

```php
/**
 * Hello World
 *
 * @api
 * @since 1.0.0
 * @param string $str Some string parameter
 */
```

The way highlight.js works (and in turn, highlight.php, since it's a strict port), you would get this HTML:

```html
<span class="hljs-comment">/**
 * Hello World
 *
 * <span class="hljs-doctag">@api</span>
 * <span class="hljs-doctag">@since</span> 1.0.0
 * <span class="hljs-doctag">@param</span> string $str Some string parameter
 */</span>
```

If you split up by newlines, you'd break `<span>`s that have new lines in them. You'd get something like:

```html
<div class="loc"><span><span class="hljs-comment">/**</span></div>
<div class="loc"><span> * Hello World</span></div>
<div class="loc"><span> *</span></div>
<div class="loc"><span> * <span class="hljs-doctag">@api</span></span></div>
<div class="loc"><span> * <span class="hljs-doctag">@since</span> 1.0.0</span></div>
<div class="loc"><span> * <span class="hljs-doctag">@param</span> string $str Some string parameter</span></div>
<div class="loc"><span> */</span></span></div>
```

Reply:
If there's a better and robust solution to splitting into lines without parsing it, I'm all ears! 😅

Reply:
Good observation. I don't have any better recommendation!